datatransfer/message/transfer_message.go

```diff
@@ -0,0 +1,106 @@
+package message
+
+import (
+	"io"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/lotus/datatransfer"
+)
+
+type transferMessage struct {
+	XferID uint64
+	IsRq   bool
+
+	Request  *transferRequest
+	Response *transferResponse
+}
+
+// ========= DataTransferMessage interface
+// Interface methods that are specific to DataTransferRequest, or to DataTransferResponse, are passthrough
+// methods to each. These methods first check whether it is a request or a response, and either call the correct method
+// on the request or response, or return zero-values as appropriate if it is the "wrong" type.
+// Consumers of DTM are expected to make a check for the right type, i.e., call IsRequest, and then make the
+// appropriate cast, however, mis-casting will not panic.  See message_test for more details about expectations and usage.
+
+// IsRequest returns true if this message is a data request
+func (tm *transferMessage) IsRequest() bool {
+	return tm.IsRq
+}
+
+// IsResponse returns true if this message is a data response
+func (tm *transferMessage) IsResponse() bool {
+	return !tm.IsRq
+}
+
+// TransferID returns the TransferID of this message
+func (tm *transferMessage) TransferID() datatransfer.TransferID {
+	return datatransfer.TransferID(tm.XferID)
+}
+
+// IsPull returns the true if this message is a Pull request, false if it is not or is not a request at all.
+func (tm *transferMessage) IsPull() bool {
+	if tm.IsRq && tm.Request != nil {
+		return tm.Request.IsPull()
+	}
+	return false
+}
+
+// Voucher returns the Voucher bytes for a DataTransferRequest. If it is not a request it returns nil always.
+func (tm *transferMessage) Voucher() []byte {
+	if tm.IsRq && tm.Request != nil {
+		return tm.Request.Voucher()
+	}
+	return nil
+}
+
+// VoucherType returns the Voucher type for a DataTransferRequest. If it is not a request or request is nil
+// it returns "invalid"
+func (tm *transferMessage) VoucherType() string {
+	if tm.IsRq && tm.Request != nil {
+		return tm.Request.VoucherType()
+	}
+	return "invalid"
```

Comment:
Once voucher types are determined this will be removed and added to a list of types.

Reply:
voucher types are never going to be determined -- the idea is that the DT module caller can add new types by registering them. If the otherside doesn't have them registered, then it will need to error. However, that will happen above this code -- in the rest of the DT module.